src/input-states.js

```diff
@@ -0,0 +1,289 @@
+ *  inputState(component, vnid, portName, state) : this
+ *  TODO inputState(component, vnid, addressablePortName, [state]) : this
+ *  TODO inputState(component, vnid, addressablePortName, socketIndex, state) : this
+ *  TODO inputState(component, vnid, addressablePortName, sourceNodeName, sourcePortName, state) : this
+ */ 
+module.exports = function(component, vnid, inportName, sourceNodeName, sourcePortName, newState) {
+    if (!component.inPorts || !component.outPorts) throw Error("This isn't a Component");
+    else if (!_.isString(vnid)) throw Error("Invalid vnid: " + vnid);
+    else if (_.isString(inportName) && !component.inPorts[inportName])
+        throw Error("Invalid port name: " + inportName);
+    var delegate = chooseForm(arguments);
+    return delegate.apply(this, arguments);
+}
+
+/**
+ * Higher level function to choose which dispatch function should be called.
+ * @param component a Component or a component facade
+ * @param args the arguments to inputState function call
+ */
+function chooseForm(args){
+    var string = _.isString(args[2]);
+    var addressable = string && args[0].inPorts[args[2]].isAddressable();
+    var finite = _.isFinite(args[3]);
+    var array = _.isArray(args[3]);
+    switch (args.length) {
+        case 2: return getAllPortStates;
+        case 3: return addressable ? getPortStateArray :
+            string ? getPortState : setAllPortStates;
+        case 4: return array ? setPortStateArray :
+            finite ? getPortStateIndex : setPortState;
+        case 5: return finite ? setPortStateIndex : getPortStateEntry;
+        case 6: return setPortStateEntry;
+        default: throw Error("Unknown set of " + args.length + " args");
+    }
+}
+
+/**
+ * Returns a hash of port names to port states or array of states if the port is
+ * addressable. This provides the caller with all portNames, distingishes between
+ * addressable and non-addressable ports, and provides the caller with all their
+ * states that will be useful when calling an updater.
+ * @param component a Component or a component facade
+ * @param vnid an identifier that distinguishes the set of states
+ */
+function getAllPortStates(component, vnid) {
+    var inPorts = _.pick(component.inPorts, isPort);
+    return _.mapObject(inPorts, function(port, portName){
+        return getPortState(component, vnid, portName);
+    });
+}
+
+/**
+ * Returns the state of the port or an array of states, if the port is
+ * addressable. When all the port states are not need this is a convient way to
+ * retrieve just the port state that is needed.
+ * @param component a Component or a component facade
+ * @param vnid an identifier that distinguishes the set of states
+ * @param portName the InPort name on this component 
+ */
+function getPortState(component, vnid, portName) {
+    var port = component.inPorts[portName];
+    if (port.isAddressable()) return getPortStateArray(component, vnid, portName);
```

Comment:
Let's use the term 'multi' instead of 'addressable' in Framework code.  The only place 'addressable' should be used would be in the noflo access layer.